File: lr/src/questions/q_1143_longest_common_subsequence.rs
pub fn longest_common_subsequence(text1: String, text2: String) -> i32 {
    let mut dp = vec![vec![0; text2.len() + 1]; text1.len() + 1];

    for i in (0..=text1.len() - 1).rev() {
        for j in (0..=text2.len() - 1).rev() {
            if text1.chars().nth(i).unwrap() == text2.chars().nth(j).unwrap() {
                dp[i][j] = 1 + dp[i + 1][j + 1];
            } else {
                dp[i][j] = i32::max(dp[i][j + 1], dp[i + 1][j]);
            }
        }
    }

    dp[0][0]
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_longest_common_subsequence() {
        assert_eq!(
            longest_common_subsequence("abcde".to_string(), "ace".to_string()),
            3
        );
        assert_eq!(
            longest_common_subsequence("abc".to_string(), "abc".to_string()),
            3
        );
        assert_eq!(
            longest_common_subsequence("abc".to_string(), "def".to_string()),
            0
        );
    }
}

File: lg/src/questions/q_1143_longest_common_subsequence.go
package questions

func longestCommonSubsequence(text1 string, text2 string) int {
	var dfs func(i, j int) int
	dfs = func(i, j int) int {
		if i >= len(text1) || j >= len(text2) {
			return 0
		}

		if text1[i] == text2[j] {
			return 1 + dfs(i+1, j+1)
		} else {
			return max(dfs(i+1, j), dfs(i, j+1))
		}
	}
	return dfs(0, 0)
}

func longestCommonSubsequenceIteration(text1 string, text2 string) int {
	m, n := len(text1), len(text2)
	dp := make([][]int, m+1)
	for i := range dp {
		dp[i] = make([]int, n+1)
	}

	for i := m - 1; i >= 0; i-- {
		for j := n - 1; j >= 0; j-- {
			if text1[i] == text2[j] {
				dp[i][j] = 1 + dp[i+1][j+1]
			} else {
				dp[i][j] = max(dp[i+1][j], dp[i][j+1])
			}
		}
	}
	return dp[0][0]
}
